### ISSUE TITLE
Unable to register key bindings (sometimes) after reload

### ISSUE THREAD

**Issue by mafredri** [kasper/phoenix#94]
Occasionally I'm struck by this when Phoenix reloads it's configuration (I can't recall if it has ever happened on initial start) and results in my key bindings not working.

```
2016-05-13 22:05:43.193 Phoenix[64300:8061225] Error: Could not register key 48 (72) with modifiers 6400. (-9878)
2016-05-13 22:05:43.194 Phoenix[64300:8061225] Error: Could not register key 123 (73) with modifiers 6400. (-9878)
2016-05-13 22:05:43.194 Phoenix[64300:8061225] Error: Could not register key 124 (74) with modifiers 6400. (-9878)
2016-05-13 22:05:43.194 Phoenix[64300:8061225] Error: Could not register key 126 (75) with modifiers 6400. (-9878)
2016-05-13 22:05:43.195 Phoenix[64300:8061225] Error: Could not register key 125 (76) with modifiers 6400. (-9878)
2016-05-13 22:05:43.195 Phoenix[64300:8061225] Error: Could not register key 36 (77) with modifiers 6400. (-9878)
2016-05-13 22:05:43.195 Phoenix[64300:8061225] Error: Could not register key 46 (78) with modifiers 6400. (-9878)
2016-05-13 22:05:43.195 Phoenix[64300:8061225] Error: Could not register key 27 (79) with modifiers 6400. (-9878)
2016-05-13 22:05:43.195 Phoenix[64300:8061225] Error: Could not register key 44 (80) with modifiers 6400. (-9878)
```

- Version: 2.1.1
- OS X: 10.11.4

**Comment by kasper** [comment]
Maybe a duplicate of #83. When this happens, are you saving your configuration especially frequently?

**Comment by mafredri** [comment]
I'm not sure, it is auto-compiled on save so the chances are that sometimes it happens quite frequently. But it does sounds pretty much as described in #83.

I can reproduce this quite easily even with a 5 second delay (`while :; do print touching; touch out/phoenix.js && sleep 5; done`).

I also tried to following to see if I can try to re-enable them later:

```javascript
while (true) {
	let all = keyHandlers.map((h) => {
		let enabled =  h.isEnabled();
		log(JSON.stringify(h.modifiers), h.key, enabled);
		if (!enabled) {
			enabled = h.enable();
			log(enabled);
		}
		return enabled;
	});
	if (all.every(x => x)) {
		break;
	}
}
```

But then I'm stuck in an eternal loop like this:

```
2016-05-14 11:13:43.095 Phoenix[77852:8155840] ["cmd","ctrl","alt"] tab 0
2016-05-14 11:13:43.095 Phoenix[77852:8155840] Error: Could not register key 48 (81) with modifiers 6400. (-9878)
2016-05-14 11:13:43.095 Phoenix[77852:8155840] 0
2016-05-14 11:13:43.095 Phoenix[77852:8155840] ["cmd","ctrl","alt"] left 0
2016-05-14 11:13:43.095 Phoenix[77852:8155840] Error: Could not register key 123 (82) with modifiers 6400. (-9878)
2016-05-14 11:13:43.095 Phoenix[77852:8155840] 0
2016-05-14 11:13:43.095 Phoenix[77852:8155840] ["cmd","ctrl","alt"] right 0
2016-05-14 11:13:43.095 Phoenix[77852:8155840] Error: Could not register key 124 (83) with modifiers 6400. (-9878)
2016-05-14 11:13:43.095 Phoenix[77852:8155840] 0
...
```

Then I thought of trying to do this without occupying the main thread:

```typescript
reallyEnableAllKeys(keyHandlers);

function reallyEnableAllKeys(kh: KeyHandler[]) {
	let ok = kh.every(h => !h.isEnabled() ? h.enable() : h.isEnabled());
	if (!ok) {
		log('All keys not enabled, retrying in 100 ms...')
		setTimeout(() => reallyEnableAllKeys(kh), 100);
	}
}
```

And voilà, they got rebound :).

---

I'm not sure how well this can be improved on the Phoenix side, but it's a bit concerning that this could happen to essentially anybody modifying their Phoenix config, and if they aren't monitoring the log output, they would have no way of knowing why their keybindings aren't working.

**Comment by kasper** [comment]
Yes, I think the main culprit is that JavaScriptCore memory management is delayed. In some instances when the context is cleared, it seems the keys are not unbound on time. Maybe we need to specifically unbind every key in the reset to ensure the state.

**Comment by mafredri** [comment]
That's what I suspected as well. Would it be easy supporting such a feature considering that we do not know where the user is storing the key handlers? Is it necessary to check all objects or could we store some sort of weak references to the handlers in Phoenix?

**Comment by kasper** [comment]
All other handlers shouldn’t cause similar side effects. Phoenix in fact already stores weak references to key handlers, so asking each handler to unbind is easy.

**Comment by mafredri** [comment]
Ah, I was only referring to the key handlers, but was perhaps a bit ambiguous in my description.

> Phoenix in fact already stores weak references to key handlers, so asking each handler to unbind is easy.

Oh, that's great! 👍

**Comment by kasper** [comment]
This should now finally be fixed in master and will be released in 2.1.2.

**Comment by mafredri** [comment]
Awesome, thanks! :)
